### ISSUE TITLE
NaN in Parquet stats (again)

### ISSUE THREAD

**Issue by crepererum** [apache/arrow-rs#264]
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
After #256 we completely ignore NaNs in parquet statistics. However, there are good reasons to fully order floats and include NaN somewhere:

- Databases usually need it, see https://github.com/cockroachdb/cockroach/issues/18860 and if DataFusion wants to follow the Postgres dialect we will have the same issue there
- Other parquet users are interested as well: [PARQUET-1222](https://issues.apache.org/jira/browse/PARQUET-1222)
- Just ignoring NaN is kinda having an implicit NULL for float while we already have a NULL-mask

**Describe the solution you'd like**
Put NaNs at the end of the float scale, so the order is:

1. -inf
2. "ordinary" numbers
3. +inf
4. NaN

**Describe alternatives you've considered**
Keeping the status quo (aka current master).

**Additional context**
.

**Comment by NGA-TRAN** [comment]
I think a few links to other DBs related documents will help us feel more confident that in the future our customers won't ask us to do differently. A few name: SQL standard (SQL99, ...),  PG, Oracle, Snowflake, Vertica.

**Comment by crepererum** [comment]
# Prior Art

## PostgreSQL
> In most implementations of the “not-a-number” concept, NaN is not considered equal to any other numeric value (including NaN). In order to allow numeric values to be sorted and used in tree-based indexes, PostgreSQL treats NaN values as equal, and greater than all non-NaN values.

https://www.postgresql.org/docs/13/datatype-numeric.html

**=> Follows this proposal**

## CockroachDB
Follows PostgreSQL, see https://github.com/cockroachdb/cockroach/issues/18860

**=> Follows this proposal**

## Oracle
> The floating-point value NaN is greater than any other numeric value and is equal to itself.

https://docs.oracle.com/database/121/TTSQL/types.htm#TTSQL165

**=> Follows this proposal**

## Snowflake
> Comparison semantics for `'NaN'` differ from the IEEE 754 standard in the following ways:
>
> Condition | Snowflake | IEEE 754 | Comment
> -- | -- | -- | --
> `'NaN' = 'NaN'` | `TRUE` | `FALSE` | In Snowflake, `'NaN'` values are all equal.
> `'NaN' > X`  where `X` is any FLOAT value, including  infinity (other than `NaN` itself). | `TRUE` | `FALSE` | Snowflake treats `'NaN'` as greater  than any other FLOAT value,  including infinity.


https://docs.snowflake.com/en/sql-reference/data-types-numeric.html#float-float4-float8

**=> Follows this proposal**

## Vertica
> Vertica follows the IEEE definition of NaNs (IEEE 754). The SQL standards do not specify how floating point works in detail. [...]
> 
> ### Rules
> -  -0 == +0
> - 1/0 = Infinity
> - 0/0 == Nan
> - NaN != anything (even NaN)
>
> [...]
>
> ### Sort Order (Ascending)
> - NaN
> - -Inf
> - numbers
> - +Inf
> - NULL

https://www.vertica.com/docs/9.2.x/HTML/Content/Authoring/SQLReferenceManual/DataTypes/Numeric/DOUBLEPRECISIONFLOAT.htm

**=> Does NOT follow this proposal**

## SQL Standard
Not open, and scanning through the drafts I've found does not reveal more than "implementation defined". I trust Vertica when they say:

> The SQL standards do not specify how floating point works in detail.

https://www.vertica.com/docs/9.2.x/HTML/Content/Authoring/SQLReferenceManual/DataTypes/Numeric/DOUBLEPRECISIONFLOAT.htm

**=> Undefined**

**Comment by alamb** [comment]
Some other potentially related prior work: https://github.com/apache/arrow/pull/9416, where @ritchie46  was trying to sort out how to handle NaNa

**Comment by jorgecarleitao** [comment]
IMO we should use the total order defined in arrow according to `IEEE 754 totalOrder`, which AFAIK is also used to compute the min and max of an array.

**Comment by crepererum** [comment]
For what it's worth: With IEEE, if a database built on top of arrow / datafusion uses a dedicated NaN representation, it could archive whatever sorting it wants by picking the "right" NaN (pick a positive one and you get the PostgreSQL behavior).

**Comment by alamb** [comment]
I think keeping "Nan" as the min/max value in the stats is unlikely going to be useful for query processing, much like storing NULLs as a min/max value is not super helpful). I would personally suggest keeping them out of the statistics and adding additional metadata (like "contains NaNs" or whatever) if that is important information to retain. 

If we really want to add NaN to the stats I think it would help to articulate an actual usecase where having a NaN there would be useful

**Comment by crepererum** [comment]
> If we really want to add NaN to the stats I think it would help to articulate an actual usecase where having a NaN there would be useful

The following come to my mind:

- PostgreSQL-style SQL (which IIRC DataFusion follows) allows `X == NaN` to be queried. For that, "contains NaN" is an important information
- PostgreSQL-style SQL (again for DataFusion) sorts NaNs at the end (aka after +inf) of the spectrum, so at least "contains NaN" would be useful.

However, "contains NaN" only makes sense for Parquet Float and Double and the current parquet standard stat do NOT provide us any way to express this. And since for most query cases the point can be made that NaN shall be included into the total order somewhere, I would rather put that info into min/max instead of creating some convoluted edge case, i.e.

> Query processors rely on min/max to figure out the full value range. However for Float and Double (and only for these) there is also this special flag that does not exist for any other data type that tells you about the existence of that one value class.

**Comment by jorgecarleitao** [comment]
Think about this, this seems to be an issue present in all implementations that use the parquet format. Maybe we should raise this over the parquet mailing list? The question seems to be what do implementations expect from `min`, `max` for float types, and whether they themselves use nans, infinites, etc, or not.

Maybe this motivates a clarification on the parquet specification itself?

**Comment by alamb** [comment]
I am coming at this from a query processing point of view and using statistics to rule out entire row groups (or equivalents). Having a clearly defined sort order for floats is important for sorting, but for statistics I feel like including `Nan` values in a column effectively "poisons" the effective use

I'll copy / paste my example from https://github.com/influxdata/influxdb_iox/pull/1448#discussion_r628230059. In  that case,  where your data was like
```
f
---
1.1
2.1
... (1 Billion other values between 1.1 and 9.9)
Nan
9.9
```

If you have a predicate that is like `f > 10.0` (which does not evaluate to true for the one `Nan` row) the query engine will have to scan 1 Billion extra rows due to the presence of a single null value

Queries that are specifically looking for Nan I think are much less important to optimize.

**Comment by alamb** [comment]
I tried to see what postgres kept for statistics, and it turns out it keeps histograms not just min/max:

```
alamb=# select * from example;
  a  
-----
   1
   2
   3
 NaN
(4 rows)

alamb=# select * from pg_stats where tablename = 'example';
 schemaname | tablename | attname | inherited | null_frac | avg_width | n_distinct | most_common_vals | most_common_freqs | histogram_bounds | correlation | most_common_elems | most_common_elem_freqs | elem_count_histogram 
------------+-----------+---------+-----------+-----------+-----------+------------+------------------+-------------------+------------------+-------------+-------------------+------------------------+----------------------
 public     | example   | a       | f         |         0 |         8 |         -1 |                  |                   | {1,2,3,NaN}      |           1 |                   |                        | 
(1 row)
```

**Comment by emkornfield** [comment]
I believe there is an open issue for Parquet to define consistent ordering here, I'll see if I can dig it up tonight.

**Comment by nevi-me** [comment]
> * Other parquet users are interested as well: [PARQUET-1222](https://issues.apache.org/jira/browse/PARQUET-1222)

@emkornfield is it this one?

**Comment by crepererum** [comment]
I'm a bit low on resources this week. I can certainly write down the options (which are more than the proposal above) and how the link to use cases and prior art next week and post it to the ticket.

**Comment by emkornfield** [comment]
Yes, sorry, PARQUET-1222 is the one and I see it was already mentioned.

**Comment by Jefffrey** [comment]
I think this should be good to close now?

PARQUET-1222 has been merged (https://github.com/apache/parquet-format/pull/185) a while back and looks like `parquet` crate is already supporting this new defined behaviour, e.g.

https://github.com/apache/arrow-rs/blob/91acfb07a9929a2d6721c5417e47c0c472372a86/parquet/src/column/writer/mod.rs#L2128-L2138

There's PARQUET-2249 which seems related but might be better off as separate issue if/when it gets merged

cc @tustvold @alamb

**Comment by crepererum** [comment]
The format now clearly states:

> NaNs should not be written to min or max statistics fields.

So parquet-rs is currently standard complaint and we should NOT diverge from the standard.

An extra counter field is currently proposed here: https://github.com/apache/parquet-format/pull/196

When this is merged I think we should implement this, but I think in the meantime we can close this ticket to make it clear that we don't plan to take any action (apart from following the upstream standard of course).

**Comment by tustvold** [comment]
Agreed, I intend to write up the proposal on https://github.com/apache/parquet-format/pull/196 to define a TypeDefinedOrder using the totalOrder predicate, as I think that would be the cleanest way to bring an end to this entire category of problem
